### ISSUE TITLE
`num_partitions` does not work for GPU

### ISSUE THREAD

**Issue by Namco0816** [google-research/t5x#410]
Hi thanks for the great work.
I were already carefully read the docs of the partitioning, but I am still confused about how it works and what did the partitioning rules means.
I tried to run the pertaining code on a single node with 8-A100 GPU. When I pretrain the T5 with the huggingface trainer and deepspeed Zero-2, it works well. However I tried to run the pretrain with the scripts provided in the examples with

```
partitioning.PjitPartitioner:
  num_partitions = 1
  logical_axis_rules= @partitioning.standard_logical_axis_rules()

partitioning.standard_logical_axis_rules:
  activation_partitioning_dims = 2
  parameter_partitioning_dims = 2
```
, 

I get the following errors:

```
56   │ Traceback (most recent call last):
  57   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/runpy.py", line 196, in _run_module_as_main
  58   │     return _run_code(code, main_globals, None,
  59   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/runpy.py", line 86, in _run_code
  60   │     exec(code, run_globals)
  61   │   File "/mnt/cache/namco/t5x/t5x/train.py", line 659, in <module>
  62   │     gin_utils.run(main)
  63   │   File "/mnt/cache/namco/t5x/t5x/gin_utils.py", line 105, in run
  64   │     app.run(
  65   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/absl/app.py", line 312, in run
  66   │     _run_main(main, args)
  67   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/absl/app.py", line 258, in _run_main
  68   │     sys.exit(main(argv))
  69   │   File "/mnt/cache/namco/t5x/t5x/train.py", line 637, in main
  70   │     _main(argv)
  71   │   File "/mnt/cache/namco/t5x/t5x/train.py", line 657, in _main
  72   │     train_using_gin()
  73   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/gin/config.py", line 1605, in gin_wrapper
  74   │     utils.augment_exception_message_and_reraise(e, err_str)
  75   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/gin/utils.py", line 41, in augment_exception_message_and_reraise
  76   │     raise proxy.with_traceback(exception.__traceback__) from None
  77   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/gin/config.py", line 1582, in gin_wrapper
  78   │     return fn(*new_args, **new_kwargs)
  79   │   File "/mnt/cache/namco/t5x/t5x/train.py", line 321, in train
  80   │     train_state = train_state_initializer.from_checkpoint_or_scratch(
  81   │   File "/mnt/cache/namco/t5x/t5x/utils.py", line 523, in from_checkpoint_or_scratch
  82   │     or self.from_scratch(init_rng))
  83   │   File "/mnt/cache/namco/t5x/t5x/utils.py", line 395, in from_scratch
  84   │     return p_initialize_train_state_fn(init_rng)
  85   │   File "/mnt/cache/namco/t5x/t5x/partitioning.py", line 729, in __call__
  86   │     return self._pjitted_fn(*args)
  87   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/jax/experimental/pjit.py", line 267, in wrapped
  88   │     args_flat, _, params, _, out_tree, _ = infer_params(*args, **kwargs)
  89   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/jax/experimental/pjit.py", line 246, in infer_params
  90   │     jaxpr, canonicalized_out_axis_resources_flat = _pjit_jaxpr(
  91   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/jax/linear_util.py", line 272, in memoized_fun
  92   │     ans = call(fun, *args)
  93   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/jax/experimental/pjit.py", line 411, in _pjit_jaxpr
  94   │     _check_shapes_against_resources("pjit outputs", mesh.is_multi_process, mesh.shape,
  95   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/jax/experimental/pjit.py", line 588, in _check_shapes_against_resources
  96   │     raise ValueError(f"One of {what} was given the resource assignment "
  97   │ ValueError: One of pjit outputs was given the resource assignment of PartitionSpec('model', None), which implies that the size of its dimension 0 should be divisib
       │ le by 8, but it is equal to 12
  98   │   In call to configurable 'train' (<function train at 0x7f598523c790>)
```

Could you please help me to fix this error?

**Comment by adarob** [comment]
Thanks for reporting this -- it looks like a GPU-specific bug in T5X. Can you try setting 

```
partitioning.PjitPartitioner:
  model_parallel_submesh = (1, 1)
  logical_axis_rules= @partitioning.standard_logical_axis_rules()
```

**Comment by adarob** [comment]
@jekbradbury as FYI

Also, I think your activation/parameter partitioning dims options will be no-ops since you're not using model parallelism.

**Comment by Namco0816** [comment]
Thanks for your reply!
I've modified the code with:
partitioning.PjitPartitioner:
```
  model_parallel_submesh = (1, 1)
  logical_axis_rules= @partitioning.standard_logical_axis_rules()
```
However I still get the error:

```
Traceback (most recent call last):
2296   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/runpy.py", line 196, in _run_module_as_main
2297   │     return _run_code(code, main_globals, None,
2298   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/runpy.py", line 86, in _run_code
2299   │     exec(code, run_globals)
2300   │   File "/mnt/cache/namco/t5x/t5x/train.py", line 659, in <module>
2301   │     gin_utils.run(main)
2302   │   File "/mnt/cache/namco/t5x/t5x/gin_utils.py", line 105, in run
2303   │     app.run(
2304   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/absl/app.py", line 312, in run
2305   │     _run_main(main, args)
2306   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/absl/app.py", line 258, in _run_main
2307   │     sys.exit(main(argv))
2308   │   File "/mnt/cache/namco/t5x/t5x/train.py", line 637, in main
2309   │     _main(argv)
2310   │   File "/mnt/cache/namco/t5x/t5x/train.py", line 657, in _main
2311   │     train_using_gin()
2312   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/gin/config.py", line 1605, in gin_wrapper
2313   │     utils.augment_exception_message_and_reraise(e, err_str)
2314   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/gin/utils.py", line 41, in augment_exception_message_and_reraise
2315   │     raise proxy.with_traceback(exception.__traceback__) from None
2316   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/gin/config.py", line 1582, in gin_wrapper
2317   │     return fn(*new_args, **new_kwargs)
2318   │   File "/mnt/cache/namco/t5x/t5x/train.py", line 507, in train
2319   │     trainer.compile_train(first_batch)
2320   │   File "/mnt/cache/namco/t5x/t5x/trainer.py", line 549, in compile_train
2321   │     self._compiled_train_step = self._partitioner.compile(
2322   │   File "/mnt/cache/namco/t5x/t5x/partitioning.py", line 779, in compile
2323   │     return partitioned_fn.lower(*args).compile()
2324   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/jax/_src/stages.py", line 174, in compile
2325   │     self._lowering.compile(), self.in_tree, self.in_avals,
2326   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/jax/interpreters/pxla.py", line 2280, in compile
2327   │     self._executable = MeshExecutable.from_hlo(
2328   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/jax/interpreters/pxla.py", line 2371, in from_hlo
2329   │     xla_executable = dispatch.compile_or_get_cached(backend, computation, compile_options)
2330   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/jax/_src/dispatch.py", line 583, in compile_or_get_cached
2331   │     return backend_compile(backend, computation, compile_options)
2332   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/jax/_src/profiler.py", line 206, in wrapper
2333   │     return func(*args, **kwargs)
2334   │   File "/mnt/cache/namco/anaconda3/envs/jax/lib/python3.10/site-packages/jax/_src/dispatch.py", line 537, in backend_compile
2335   │     return backend.compile(built_c, compile_options=options)
2336   │ RuntimeError: UNIMPLEMENTED: Requested AllReduce not implemented on GPU; replica_count: 1; partition_count: 8, group_mode: kCrossReplicaAndPartition, operand_count
       │ : 26; NCCL support: 1; first operand array element-type: BF16
2337   │   In call to configurable 'train' (<function train at 0x7f3d01a81240>)
2338   │ Rewritten gin arg: --gin_bindings=MODEL_DIR = "/mnt/lustre/namco/jax-model/t5-base"
```

I am also very confused about the partition rules in the partitioning.py. I've noticed that the `get_gpu_mesh` will return the mesh for gpu. This function will return a (1, 8) mesh for my 8 A100 GPUs machine. The first dimension 1 represents the host_num and second dimension 8 represent the num_gpus. If I understand correctly, based on the partition rules, 'data' will be assigned to the first axis and 'model' will be assigned to the second axis.  However the simple ddp will shard the input_data across 8 GPUs, which means that the partition rules should be ("batch", "model"), however the code provided in the partitioning example define the data parallel as ("batch", "data"). I am really confused about this part.

Thanks for your help!

**Comment by ibulu** [comment]
I think I am getting a related error when trying to fine-tune longT5 model on CPU:

ValueError: Failed to map logical axes for target/decoder/logits_dense/kernel
  In call to configurable 'train' (<function train at 0x17a77b940>)

**Comment by sudhakarsingh27** [comment]
#### TLDR; 
@adarob For data+model parallelism on GPUs, is `model_parallel_submesh=(1,1,1,<#GPU for model parallelism>)` is the way to go for a single node multi-gpu case (seems to be suggested by [this line in the code](https://github.com/google-research/t5x/blob/feb97e77e885ec35a185d9a35d9a028384c7bbe1/t5x/partitioning.py#L270) as well)? 
Thanks!

---
#### More context:

I'm working with this config: [t5_1_1/base.gin](https://github.com/google-research/t5x/blob/main/t5x/examples/t5/t5_1_1/base.gin#L46) and I faced the same error as OP when I used the default partitioning config. (My intention is to run data+model parallel).

I followed @adarob 's suggestion but couldn't get it running.

> Thanks for reporting this -- it looks like a GPU-specific bug in T5X. Can you try setting
> 
> ```
> partitioning.PjitPartitioner:
>   model_parallel_submesh = (1, 1)
>   logical_axis_rules= @partitioning.standard_logical_axis_rules()
> ```

After playing around, I found the following partitioning rule:
```
partitioning.PjitPartitioner:
  model_parallel_submesh = (1, 1, 1, 2)
  logical_axis_rules= @partitioning.standard_logical_axis_rules()
```

seems to get the following mesh of devices `[2,4]` (and mapping to `('data', 'model')`) which is what I wanted.
```
[[GpuDevice(id=0, process_index=0) GpuDevice(id=1, process_index=0)]
 [GpuDevice(id=2, process_index=0) GpuDevice(id=3, process_index=0)]
 [GpuDevice(id=4, process_index=0) GpuDevice(id=5, process_index=0)]
 [GpuDevice(id=6, process_index=0) GpuDevice(id=7, process_index=0)]]
```

**Comment by sudhakarsingh27** [comment]
Also, as @Namco0816 pointed out, with using `data parallelism only`, the GPU mesh returned is 
```
[[GpuDevice(id=0, process_index=0) GpuDevice(id=1, process_index=0)                                                                                                                                        
  GpuDevice(id=2, process_index=0) GpuDevice(id=3, process_index=0)                                                                                                                                        
  GpuDevice(id=4, process_index=0) GpuDevice(id=5, process_index=0)                                                                                                                                        
  GpuDevice(id=6, process_index=0) GpuDevice(id=7, process_index=0)]]
```
which is of dim `[1,8]` and maps to `('data','model')` axes and so there's effectively no data parallelism even when [data only parallelism](https://github.com/google-research/t5x/blob/6b4cf7c7d9de80135bce14787dd1616653c99eae/t5x/partitioning.py#L286) is selected. I think [this function](https://github.com/google-research/t5x/blob/6b4cf7c7d9de80135bce14787dd1616653c99eae/t5x/partitioning.py#L286) is completely agnostic of `data`/`model` parallel axes/partitions and therefore we see the issue. 
Can someone confirm this? @adarob @jekbradbury

**Comment by adarob** [comment]
Fixed by #643
